services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateManager.java

```diff
@@ -16,73 +16,173 @@
+        Set<String> subnetIdsSet = hostIpSubnetIdsMap.get(internalPortEntity.getBindingHostIP());
+        if (subnetIdsSet == null) subnetIdsSet = new TreeSet<>();
+        fixedIpsSet.add(ip.getIpAddress());
+        subnetIdsSet.add(ip.getSubnetId());
+        hostIpFixedIpsMap.put(internalPortEntity.getBindingHostIP(), fixedIpsSet);
+        hostIpSubnetIdsMap.put(internalPortEntity.getBindingHostIP(), subnetIdsSet);
+
+        ipPortIdMap.put(ip.getIpAddress(), internalPortEntity.getId());
+        ipHostIpMap.put(ip.getIpAddress(), internalPortEntity.getBindingHostIP());
+        ipSubnetIdMap.put(ip.getIpAddress(), ip.getSubnetId());
+        ipMacMap.put(ip.getIpAddress(), internalPortEntity.getMacAddress());
+      }
+    }
+    // for neighbor infos
+    if (neighborInfos != null) {
+      for (NeighborInfo internalPortEntity : neighborInfos) {
+        ipPortIdMap.put(internalPortEntity.getPortIp(), internalPortEntity.getPortId());
+        ipHostIpMap.put(internalPortEntity.getPortIp(), internalPortEntity.getHostIp());
+        portIdNeighborInfoMap.put(internalPortEntity.getPortIp(), internalPortEntity);
+        String portIp = internalPortEntity.getPortIp();
+        Set<String> fixedIps = hostIpFixedIpsMap.get(internalPortEntity.getHostIp());
+        if (fixedIps == null) fixedIps = new TreeSet<>();
+        Set<String> subnetIds = hostIpSubnetIdsMap.get(internalPortEntity.getHostIp());
+        if (subnetIds == null) subnetIds = new TreeSet<>();
+        fixedIps.add(portIp);
+        subnetIds.add(internalPortEntity.getSubnetId());
+        hostIpFixedIpsMap.put(internalPortEntity.getHostIp(), fixedIps);
+        hostIpSubnetIdsMap.put(internalPortEntity.getHostIp(), subnetIds);
+        ipPortIdMap.put(portIp, internalPortEntity.getPortId());
+        ipHostIpMap.put(portIp, internalPortEntity.getHostIp());
+        ipSubnetIdMap.put(portIp, internalPortEntity.getSubnetId());
+        ipMacMap.put(portIp, internalPortEntity.getPortMac());
+      }
+    }
+  }
+  /**
+   * print dpm input msg
+   *
+   * @param networkConfiguration msg
+   */
+  private void printNetworkConfiguration(NetworkConfiguration networkConfiguration) {
+    ObjectMapper objectMapper = new ObjectMapper();
+    try {
+      String json = objectMapper.writeValueAsString(networkConfiguration);
+      LOG.log(Level.INFO, "@@@input json str: " + json);
+
+    } catch (JsonProcessingException e) {
+      LOG.log(Level.SEVERE, e.getMessage());
+    }
   }
   /**
    * transform client of dpm msg to aca protobuf format
    *
-   * @param networkConfiguration  msg to be transformmed
+   * @param networkConfiguration msg to be transformmed
    * @return Map<String, Goalstate.GoalState>
    * @throws RuntimeException Various exceptions that may occur during the send process
    */
-  public Map<String, Goalstate.GoalState> transformNorthToSouth(
+  @Async
+  public Future<Map<String, Goalstate.GoalState>> transformNorthToSouth(
```

Comment:
Based on the requirement listed on Issue #448, this transformNorthToSouth method has more than 500+ lines of codes and needs major refactor. The refactor so far doesn't break down the method to improve code readability and reduce potential risk of NPE.

Reply:
changed in 7ccbc91

Reply:
The major logic of transformNorthToSouth method has been moved to GoalStateTransformer, and with some logics pulled out, which is a positive move.

However, I have to point out that the major complexity remains in GoalStateTransformer, and the code remains long and prone to error or NPE (see a potential issue in line 258 of GoalStateTransformer as pointed out there).

Reply:
per discussion , we would leave this later and merge for now since logic change needs contract change